### PR TITLE
Fix do-while flow compilation

### PR DESCRIPTION
**Pull request by PerfectLaugh**
This fixes incorrect behavior of do-while which causing infinite loop when encountering continue statement.

### CODE REVIEW

**Comment by dvander** [comment]
Thanks for the fix - can you include a test case?

**Comment by PerfectLaugh** [comment]
Before:
image

After:
image

**Comment by Fyren** [comment]
I think the test for `do` always needs to be emitted, regardless of `IsTerminal`.  If not, a `continue` in the loop body will just end the loop.

```
int x = 0; 
do { 
  x++;
  continue; 
} while (x < 3);
```

**Comment by PerfectLaugh** [comment]
> I think the test for `do` always needs to be emitted, regardless of `IsTerminal`. If not, a `continue` in the loop body will just end the loop.
> 
> ```
> int x = 0; 
> do { 
>   x++;
>   continue; 
> } while (x < 3);
> ```

Thanks for your kindly correction. I didn't know about what IsTerminal means so I made the mistake.

File: tests/basic/while.sp [hunk 1]
```diff
@@ -10,4 +10,31 @@ public main()
...
+        break;
+      }
+      else {
+        print("even\n");
+      }
+
+      continue;
+    }
+
+    print("old\n");
```
Comment (KyleSanderson) [inline tests/basic/while.sp:30]: Should this say odd?

Comment (dvander) [reply tests/basic/while.sp:30]: I'm holding off merging until I can run it against the corpus, which will probably be a day or two.

If PerfectLaugh wants to fix the typo before then, it'd be good to also split these two new cases into separate test files. Debugging tests is much easier when they're small, so we try not to make a single test function too big.

But I'm okay to merge this either way, typo fixed and tests split or not, I can fix that later.

**Review by dvander — APPROVED** [review]

File: tests/basic/while.sp (hunk 1 above)
Comment (PerfectLaugh) [reply tests/basic/while.sp]: Sorry my English is bad ;\

@dvander I migrated the do-while part into separate file named "do-while.sp"

**Comment by dvander** [comment]
Thanks! I've manually merged this to master, with a few tiny fixes (split up the test further, and retained a codegen optimization where we don't have to emit on break/returns).

**Pull request closed without merging**